runtime/precompiles/src/lib.rs

```diff
@@ -16,16 +16,190 @@
+///
+/// EXAMPLE USECASE:
+/// A simple example usecase is a contract that allows donors to donate, and stakes all the funds
+/// toward one fixed address chosen by the deployer.
+/// Such a contract could be deployed by a collator candidate, and the deploy address distributed to
+/// supporters who want to donate toward a perpetual nomination fund.
+pub struct NominateWrapper<Runtime> {
+	_phantom_data: PhantomData<(Runtime)>,
+}
+
+// type BalanceOf<Runtime> =
+// 	<<Runtime as parachain_staking::Config>::Currency as Currency<Runtime::AccountId>>::Balance;
+
+impl<Runtime> Precompile for NominateWrapper<Runtime>
+where
+	Runtime: parachain_staking::Config + pallet_evm::Config,
+	Runtime::AccountId: From<H160>,
+	<<Runtime as parachain_staking::Config>::Currency as Currency<Runtime::AccountId>>::Balance:
+		TryFrom<U256> + Debug,
+	<Runtime as frame_system::Config>::Call:
+		Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+	<<Runtime as frame_system::Config>::Call as Dispatchable>::Origin:
+		From<Option<Runtime::AccountId>>,
+{
+	fn execute(
+		input: &[u8], //Reminder this is big-endian
+		target_gas: Option<u64>,
+		context: &Context,
+	) -> Result<(ExitSucceed, Vec<u8>, u64), ExitError> {
+		// Basic sanity checking for length
+		// https://solidity-by-example.org/primitives/
+
+		// Params are:
+		// Collator target (size 20 bytes, address)
+		// Nomination Amoutn: (size 32 bytes, U256 (for now))
+		const COLLATOR_SIZE_BYTES: usize = 20;
+		const AMOUNT_SIZE_BYTES: usize = 32;
+		if input.len() != COLLATOR_SIZE_BYTES + AMOUNT_SIZE_BYTES {
+			return Err(ExitError::Other(
+				"input length for Sacrifice must be exactly 16 bytes".into(),
+			));
+		}
+
+		// Parse input into Rust types
+		let mut collator_buf: [u8; COLLATOR_SIZE_BYTES] = [0; COLLATOR_SIZE_BYTES];
+		let mut amount_buf: [u8; AMOUNT_SIZE_BYTES] = [0; AMOUNT_SIZE_BYTES];
+
+		// TODO is this copy necessary? Probably.
+		collator_buf.copy_from_slice(&input[0..COLLATOR_SIZE_BYTES]);
+		amount_buf
+			.copy_from_slice(&input[COLLATOR_SIZE_BYTES..COLLATOR_SIZE_BYTES + AMOUNT_SIZE_BYTES]);
+
+		// Convert to right data types
+		let collator = H160::from_slice(&collator_buf);
+
+		// TODO make the currency type generic
+		let amount: <<Runtime as parachain_staking::Config>::Currency as Currency<
+			Runtime::AccountId,
+		>>::Balance = match sp_core::U256::from_big_endian(&amount_buf).try_into() {
+			Ok(x) => x,
+			_ => 0.into(),
+		};
+		// let temp = BigUint::from_bytes_be(&amount_buf);
+		// let amount: <<Runtime as parachain_staking::Config>::Currency as Currency<
+		// 	<Runtime as frame_system::pallet::Config>::AccountId,
+		// >>::Balance = temp.
+
+		println!("Collator account is {:?}", collator);
+		// println!("Amount is {:?}", amount);
+
+		// Construct a call
+		let call = <Runtime as frame_system::Config>::Call::ParachainStaking(parachain_staking::Call::<Runtime>::nominate(
+			collator.into(),
+			amount,
+		));
```

Comment:
```
error[E0599]: no associated item named `ParachainStaking` found for associated type `<Runtime as frame_system::Config>::Call` in the current scope
   --> runtime/precompiles/src/lib.rs:172:55
    |
172 |         let call = <Runtime as frame_system::Config>::Call::ParachainStaking(parachain_staking::Call::<Runtime>::nominate(
    |                                                             ^^^^^^^^^^^^^^^^ associated item not found in `<Runtime as frame_system::Config>::Call`
```

It seems like `ParachainStaking` is not a valid variant in the outer Call enum. We've told the compiler that the parachain staking pallet is in our runtime by including the trait bound, but the compiler doesn't know what text is used in `construct_runtime!` and thus can't be sure the variant exists. How can I construct an outer call that I can then dispatch? Any ideas @thiolliere

Reply:
Nvm, writing it out helped me understand the problem about the variant name. And then I figured out `let outer_call: Runtime::Call = inner_call.into();`